Simulation with flow rate coefficient

heyyeahcrow opened the pull request:
The simulation was successfully run after adding the flow rate factor.
# I intentionally set the activation energy for beta and kappa to be smaller than others to control the result

However,
1. The function for calculating the flow-rate coefficient still requires some updates
2. Haven't built the for loop for different flow rates
3. I haven't added the probability of growth at this time point. It probably won't make too much change to the result

heyyeahcrow committed:
1D heatmap, probably correct kmatrix

heyyeahcrow committed:
Add files via upload

The simulation was successfully run after adding the flow rate factor.
However,
1. The function for calculating the flow-rate coefficient still requires some updates
2. Haven't built the for loop for different flow rates
3. I haven't added the probability of growth at this time point. It probably won't make too much change to the result
# I intentionally set the activation energy for beta and kappa to be smaller than others to control the result

heyyeahcrow committed:
Delete 3phases.ipynb

heyyeahcrow committed:
Delete 5phases.ipynb

heyyeahcrow committed:
Delete ArrheniusEq.jl

rkurchin commented on Energies.jl:
```diff
@@ -0,0 +1,10 @@
+G_values = [-5.92, -5.942, -5.97, -5.948, -5.937, -5.94] #Amorphous unsure
+deltaG = [G_values[j] - G_values[i] for j in 1:length(G_values), i in 1:length(G_values)]
+Ea_constants = [0.00 1.00 0.01 0.01 1.00 1.00;  #set the activation energy from amorphous to kappa and beta to be small
```
if you declare this as a `Symmetric` matrix, you'll save a little bit of memory and also guard against typos etc.

rkurchin commented on Energies.jl:
```diff
@@ -0,0 +1,10 @@
+                0.01 1.00 0.01 0.00 1.00 1.00;
+                1.00 1.00 1.00 1.00 0.00 1.00;
+                1.00 1.00 1.00 1.00 1.00 0.00]
+Ea_values = [i == j ? 0 : (deltaG[i, j] > 0 ? deltaG[i, j] + Ea_constants[i, j] : Ea_constants[i, j])
```
this logic looks right to me, but it's a bit hard to follow, so maybe add a comment clarifying what it does

rkurchin commented on newArrheniusEq.jl:
```diff
@@ -0,0 +1,70 @@
+end
+
+function arrhenius_rate_matrix(Ea_m, T)
+    k_matrix = [arrhenius_rate(Ea_m[i, j], T) for i in 1:size(Ea_m, 1), j in 1:size(Ea_m, 2)]
```
this can be done more efficiently and clearly with a broadcast, i.e. 
```julia
k_matrix = arrhenius_rate.(Ea_m, T)
```

rkurchin commented on newArrheniusEq.jl:
```diff
@@ -0,0 +1,70 @@
+    k_matrix = [arrhenius_rate(Ea_m[i, j], T) for i in 1:size(Ea_m, 1), j in 1:size(Ea_m, 2)]
+    k = copy(k_matrix)
+    # Adjust the diagonal elements
+    for i in 1:size(k_matrix, 1)
```
you don't need nested for loops here, since `k_matrix` will always be square. Just loop once as in:
```julia
for i in 1:size(k_matrix,1)
   k_matrix[i,i] = ...
end
```

rkurchin reviewed (changes requested):
See comments. Also, any time you define a function that's not not purely some local convenience thing, put it in its own file rather than directly in the notebook. It makes it easier to track in git, and is also good practice from a reproducibility perspective (i.e. you can more easily know which version of a function you used to produce a given result, etc.). Once you do that, I'll look more closely at your simulation functions as well as the ones I've left comments on here.

Also, please commit a `Project.toml` file to track package dependencies.

heyyeahcrow committed:
Adaption based on the comment of pull request

heyyeahcrow committed:
0408 Basics

1. Separate the functions from the notebooks
2. Flowrate + Temperature completed (with plot functions)
3. Updates for the rate coefficient function are still required

heyyeahcrow commented:
1. Fix the above comments.
2. Complete the 2D simulation with both Temperature and Flow rate inputs.
3. Some updates for the rate coefficient function are still required.
4. Haven't built the data structure for Energies mentioned in Issues.

rkurchin commented on Simulation.jl:
```diff
@@ -0,0 +1,50 @@
+function simulate_deposition(fcoeffs, Ea_values, T_range, num_steps, dt)
```
add a docstring that explains the function and what the inputs/outputs are

rkurchin replied:
also, I would suggest just writing this for a single T value. It should work fine to compute a range either by a custom broadcast or a list comprehension, but I think for the sake of modularity, just having the one loop over `num_steps` inside this function makes more sense

rkurchin replied:
(likewise for a single `fcoeff` vector rather than a vector of vectors)

rkurchin commented on Simulation.jl:
```diff
@@ -0,0 +1,50 @@
+            for _ in 1:num_steps
+                e = copy(existing_layers)
+                for i in 1:size(existing_layers, 1)
+                    e[i, :] = fcoeff[i] * existing_layers[i, :] # for more inner layers, multiply by smaller fcoeff
```
we should write this in a way that the `e` array doesn't change size during evaluation. Precompute what its size is an initialize accordingly, then just populate as you go.

rkurchin replied:
also, this step should be able to be done as an elementwise multiply without having to write out the loop, much easier to read.

rkurchin commented on newArrheniusEq.jl:
```diff
@@ -0,0 +1,56 @@
+"""
+
+function flow_coefficient(type, effecting_nums, decay_coefficients)
+    flow_coefficients = [Float64[] for _ in 1:length(decay_coefficients)]
```
again, initialize at the size it needs to be and then populate rather than `push!`ing to it

rkurchin commented on newArrheniusEq.jl:
```diff
@@ -0,0 +1,56 @@
+end
+"""
+
+function flow_coefficient(type, effecting_nums, decay_coefficients)
```
write this for a single value of the decay coefficient for easier readability, as I said above, can either define a custom broadcast if needed to work on an array, or just do that as a list comprehension later

rkurchin reviewed (changes requested):
left some more feedback here – the data structure should be a separate PR, that's why I put it as an issue rather than as a comment here.

heyyeahcrow committed:
Update based on the pull comments

Adapt data structure of Flow-Coefficients
Move T and Flow-rates for loops outside the simulation function
Add annotation for in/output for some functions
